[clang][scandeps] Improve handling of rawstrings.

tru opened the pull request:
The current parser just checks one step back for a R before each string to know that it's a rawstring. But the preprocessor is much more advanced here and can have constructs like:

R\
"str"

And much more. This patch also adds more test coverage for Rawstrings in the dependencydirectivesscanner.

This was co-authored by Sylvain Audi <sylvain.audi@ubisoft.com> (@sylvain-audi)

Fixes #137648

tru committed:
[clang][scandeps] Improve handling of rawstrings.

The current parser just checks one step back for a R before each string
to know that it's a rawstring. But the preprocessor is much more
advanced here and can have constructs like:

R\
"str"

And much more. This patch also adds more test coverage for Rawstrings in
the dependencydirectivesscanner.

This was co-authored by Sylvain Audi <sylvain.audi@ubisoft.com>

Fixes #137648

benlangmuir commented on clang/lib/Lex/DependencyDirectivesScanner.cpp:
```diff
@@ -215,25 +233,28 @@ static void skipOverSpaces(const char *&First, const char *const End) {
 static void skipRawString(const char *&First, const char *const End) {
   assert(First[0] == '"');
-  assert(First[-1] == 'R');
+  //assert(First[-1] == 'R');
```
Please remove this line

benlangmuir commented on clang/lib/Lex/DependencyDirectivesScanner.cpp:
```diff
@@ -206,6 +206,24 @@ static void skipOverSpaces(const char *&First, const char *const End) {
+static char previousChar(const char *First, const char *&Current) {
+  assert(Current > First);
+  --Current;
+  while (Current > First + 1 && isVerticalWhitespace(*Current)) {
```
The Lexer allows trailing horizontal whitespace after the `\` (with a warning).  Maybe we should make `Lexer::getEscapedNewLineSize` public and use that here to keep it in sync?

tru committed:
[clang][lexer] Make Lexer::getEscapedNewLineSize public

This is so that we can use it in the dependency scanner without duplicating the logic there.

tru committed:
Improvements to raw-string parsing and expanding tests

As suggested in the review, use Lexer::getEscapedNewLineSize() instead
of implementing our own logic for it.

tru committed:
Remove assertion that's no longer valid here

tru commented:
Sorry for the delay. I have addressed both your comments now @benlangmuir - reworked it to use Lexer::getEscapedNewLineSize() and removed the assertline. I also expanded the test coverage.

tru committed:
clang-format
